File: tools/netsim/src/controller/device.cc
// Copyright 2022 The Android Open Source Project
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "controller/device.h"

#include <cstdint>
#include <string>
#include <string_view>
#include <vector>

#include "common.pb.h"
#include "hci/bluetooth_facade.h"
#include "model.pb.h"
#include "netsim-cxx/src/lib.rs.h"
#include "util/log.h"
#include "wifi/wifi_facade.h"

namespace netsim {
namespace controller {

namespace {
// To detect bugs of misuse of chip_id more efficiently.
const int kGlobalChipStartIndex = 1000;
}  // namespace

model::Device Device::Get() {
  model::Device model;
  model.set_id(id);
  model.set_name(name);
  model.set_visible(visible);
  model.mutable_position()->CopyFrom(position);
  model.mutable_orientation()->CopyFrom(orientation);

  // populate chips models
  for (auto &[_, chip] : chips_) {
    model.add_chips()->CopyFrom(chip->Get());
  }
  return model;
}

void Device::Patch(const model::Device &request) {
  if (this->visible != request.visible()) {
    this->visible = request.visible();
  }
  if (request.has_position()) {
    this->position.CopyFrom(request.position());
  }
  if (request.has_orientation()) {
    this->orientation.CopyFrom(request.orientation());
  }
  for (const auto &request_chip_model : request.chips()) {
    // TODO: use request_chip_model.kind()
    auto request_chip_kind =
        request_chip_model.kind() != common::ChipKind::UNSPECIFIED
            ? request_chip_model.kind()
            : common::ChipKind::BLUETOOTH;
    const auto &request_chip_name = request_chip_model.name();
    BtsLog("request kind:%d, name:%s", request_chip_kind,
           request_chip_name.c_str());
    for (auto &[_, chip] : chips_) {
      BtsLog("chip kind:%d, name:%s", chip->kind, chip->name.c_str());
      if (chip->name == request_chip_name && chip->kind == request_chip_kind) {
        chip->Patch(request_chip_model);
      }
    }
  }
}

bool Device::RemoveChip(uint32_t chip_id) {
  if (chips_.find(chip_id) != chips_.end()) {
    BtsLog("Device::RemoveChip: removed %d", chip_id);
    chips_[chip_id]->Remove();
    chips_.erase(chip_id);
  } else {
    BtsLog("Device::RemoveChip: %d not found", chip_id);
  }
  return chips_.size() == 0;
}

std::pair<uint32_t, uint32_t> Device::AddChip(common::ChipKind chip_kind,
                                              const std::string &chip_name,
                                              const std::string &manufacturer,
                                              const std::string &product_name) {
  for (auto &[_, chip] : chips_) {
    auto chip_model = chip->Get();
    if (chip_model.kind() == chip_kind && chip_model.name() == chip_name) {
      BtsLog("Device::AddChip - duplicate, skipping.");
      return {-1, -1};
    }
  }
  static uint32_t global_chip_id = kGlobalChipStartIndex;
  auto chip_id = global_chip_id++;
  uint32_t facade_id;
  if (chip_kind == common::ChipKind::BLUETOOTH) {
    facade_id = hci::facade::Add(this->id);
    BtsLog("hci::facade::Add chip_id:%d, facade_id:%d", id, facade_id);
  } else if (chip_kind == common::ChipKind::WIFI) {
    facade_id = wifi::facade::Add(this->id);
  } else if (chip_kind == common::ChipKind::UWB) {
    facade_id = uwb::facade::Add(this->id);
  } else {
    BtsLog("Device::AdChip: unable to add chip");
    return {-1, -1};
  }
  auto chip = std::make_shared<Chip>(chip_id, facade_id, chip_kind, chip_name,
                                     this->name, manufacturer, product_name);
  chips_[chip_id] = std::move(chip);
  return {chip_id, facade_id};
}

void Device::Reset() {
  this->visible = true;
  this->position.Clear();
  this->orientation.Clear();
  for (auto &[_, chip] : chips_) {
    chip->Reset();
  }
}

void Device::Remove() {
  for (auto &[_, chip] : chips_) {
    chip->Remove();
  }
}

}  // namespace controller
}  // namespace netsim

File: tools/netsim/rust/netsim-cxx/src/devices/device.rs
// Copyright 2023 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Device.rs

use protobuf::Message;

use crate::devices::chip;
use crate::devices::chip::Chip;
use crate::devices::chip::ChipIdentifier;
use crate::devices::facades::FacadeIdentifier;
use frontend_proto::common::ChipKind as ProtoChipKind;
use frontend_proto::model::Device as ProtoDevice;
use frontend_proto::model::Orientation as ProtoOrientation;
use frontend_proto::model::Position as ProtoPosition;
use std::collections::HashMap;

pub type DeviceIdentifier = i32;

pub struct Device {
    pub id: DeviceIdentifier,
    pub guid: String,
    name: String,
    visible: bool,
    pub position: ProtoPosition,
    orientation: ProtoOrientation,
    pub chips: HashMap<ChipIdentifier, Chip>,
}
impl Device {
    pub fn new(id: DeviceIdentifier, guid: String, name: String) -> Self {
        Device {
            id,
            guid,
            name,
            visible: true,
            position: ProtoPosition::new(),
            orientation: ProtoOrientation::new(),
            chips: HashMap::new(),
        }
    }
}

pub struct AddChipResult {
    pub device_id: DeviceIdentifier,
    pub chip_id: ChipIdentifier,
    pub facade_id: FacadeIdentifier,
}

impl Device {
    pub fn get(&self) -> ProtoDevice {
        let mut device = ProtoDevice::new();
        device.id = self.id;
        device.name = self.name.clone();
        device.visible = self.visible;
        device.position = protobuf::MessageField::from(Some(self.position.clone()));
        device.orientation = protobuf::MessageField::from(Some(self.orientation.clone()));
        for chip in self.chips.values() {
            device.chips.push(chip.get());
        }
        device
    }

    /// Patch a device and its chips.
    pub fn patch(&mut self, patch: &ProtoDevice) {
        // TODO visible should be State
        self.visible = patch.visible;
        if patch.position.is_some() {
            self.position.clone_from(&patch.position);
        }
        if patch.orientation.is_some() {
            self.orientation.clone_from(&patch.orientation);
        }
        // iterate over patched ProtoChip entries and patch matching chip
        for patch_chip in patch.chips.iter() {
            // Allow default chip kind of BLUETOOTH
            let patch_chip_kind = patch_chip.kind.enum_value_or(ProtoChipKind::BLUETOOTH);
            let patch_chip_name = &patch_chip.name;
            // Find the matching chip and patch the proto chip
            for chip in self.chips.values_mut() {
                if chip.name.eq(patch_chip_name) && chip.kind == patch_chip_kind {
                    chip.patch(patch_chip);
                    break; // next proto chip
                }
            }
        }
    }

    /// Remove a chip from a device.
    pub fn remove_chip(&mut self, chip_id: ChipIdentifier) {
        if let Some(chip) = self.chips.get_mut(&chip_id) {
            chip.remove();
        } else {
            eprintln!("RemoveChip id {chip_id} not found");
        }
        self.chips.remove(&chip_id);
    }

    pub fn add_chip(
        &mut self,
        device_name: &str,
        chip_kind: ProtoChipKind,
        chip_name: &str,
        chip_manufacturer: &str,
        chip_product_name: &str,
    ) -> Option<AddChipResult> {
        for chip in self.chips.values() {
            if chip.kind == chip_kind && chip.name == chip_name {
                eprintln!("Device::AddChip - duplicate at id {}, skipping.", chip.id);
                return None;
            }
        }
        let chip = chip::chip_new(
            self.id,
            chip_kind,
            chip_name,
            device_name,
            chip_manufacturer,
            chip_product_name,
        );
        let chip_id = chip.id;
        let facade_id = chip.facade_id;
        self.chips.insert(chip.id, chip);
        Some(AddChipResult { device_id: self.id, chip_id, facade_id })
    }

    /// Reset a device to its default state.
    pub fn reset(&mut self) {
        self.visible = true;
        self.position.clear();
        self.orientation.clear();
        for chp in self.chips.values_mut() {
            chp.reset();
        }
    }

    /// Remove all chips from a device.
    /// Called at shutdown.
    #[allow(dead_code)]
    pub fn remove(&mut self) {
        for (_, chip) in self.chips.iter_mut() {
            chip.remove();
        }
    }
}
